[#795] : Remove geth/common package

adriacidre opened the pull request:
Last bits of `get/common` package has been moved to its appropriate packages.

### Important changes:

- [x] Moved all transaction related logic to geth/transactions
- [x] Moved all testing specific logic to t/utils package
- [x] Renamed QueueTxID to QueuedID
- [x] Moved specific interfaces as CompleteTransactionResult to `lib/`
- [x] Moved conversion between `Call` and `SendTxArgs` to transactions package to avoid crossed depenndencies


Closes #795

adriacidre committed:
[#795] : Remove geth/common package

mandrigin commented on cmd/statusd/debug/commands.go:
```diff
@@ -170,7 +170,7 @@ func (cs *commandSet) Logout() error {
 // CompleteTransaction instructs API to complete sending of a given transaction.
 func (cs *commandSet) CompleteTransaction(id, password string) (string, error) {
-	txHash, err := cs.statusAPI.CompleteTransaction(common.QueuedTxID(id), password)
+	txHash, err := cs.statusAPI.CompleteTransaction(utils.QueuedTxID(id), password)
```
Maybe `transactions.QueuedID` is a better name?

mandrigin commented on geth/api/api.go:
```diff
@@ -147,27 +147,27 @@ func (api *StatusAPI) Logout() error {
 
 // SendTransaction creates a new transaction and waits until it's complete.
-func (api *StatusAPI) SendTransaction(ctx context.Context, args common.SendTxArgs) (gethcommon.Hash, error) {
+func (api *StatusAPI) SendTransaction(ctx context.Context, args utils.SendTxArgs) (gethcommon.Hash, error) {
```
I would move it to `transactions` package to have `transactions.SendTxArgs`. I'm not sure that `SendTxArgs` is a good name either, but let's keep it for another time.

adriacidre replied:
Yep, previously I had the limitation of cyclic dependency, thats why it lives on utils. However as discussed with @adambabik I'll try to merge everything under `geth/transactions` package.

adambabik reviewed (commented):
`transactions/utils` package seems very artificial. Maybe we should merge `transactions/queue` with `transactions` instead? We can avoid `transactions/utils` which in `queue.go` is used only to import a type.

mandrigin reviewed (approved):
Looks good!

adriacidre commented:
@adambabik yes, I can try merging `transactions/queue` and `transactions`. That will also avoid the _cyclic dependency_

adambabik commented on geth/rpc/call.go:
```diff
@@ -146,7 +146,7 @@ func (c Call) ParseGasPrice() *hexutil.Big {
 
 // ToSendTxArgs converts Call to SendTxArgs.
-func (c Call) ToSendTxArgs() common.SendTxArgs {
+func (c Call) ToSendTxArgs() utils.SendTxArgs {
```
Maybe this conversion between `Call` and `SendTxArgs` should stay on `geth` level? We couple `geth/rpc` and `geth/transactions` just because of this one method.

adambabik commented on geth/transactions/utils/utils.go:
```diff
@@ -1,4 +1,4 @@
-package common
+package utils
```
This file contains methods used in test files, for example, `ImportTestAccount` and in regular files. It would be beneficial to split it.

adambabik commented on lib/library.go:
```diff
@@ -197,15 +197,15 @@ func Logout() *C.char {
 	}
 
-	out := common.CompleteTransactionResult{
+	out := utils.CompleteTransactionResult{
```
I am not sure if `CompleteTransactionResult` should be declared in `geth/transactions`. It's used only in `lib` package, right?

adriacidre commented:
all changes have been applied, so changes at this point are:

- [x] Moved all transaction related logic to `geth/transactions`
- [x] Moved all testing specific logic to t/utils package
- [x] Renamed QueueTxID to QueuedID
- [x] Moved specific interfaces as CompleteTransactionResult to `lib/`
- [x] Moved conversion between `Call` and `SendTxArgs` to transactions package to avoid crossed depenndencies

dshulyak commented:
maybe it is worth changing QueueID to string? I wanted to do that when refactored transaction manager a bit and then forgot...

adriacidre commented:
@dshulyak so, instead of using `var ids []transactions.QueuedID` everywhere, reduce it to simply use `var ids []string`?

dshulyak commented:
@adriacidre yes, it looks like that the custom type for tx id doesn't do anything useful

adriacidre commented:
nice! using string instead of `QueuedID` reduced some dependencies on transactions package and removed all conversions from QueuedID to string 🎉

adambabik commented on geth/transactions/queue.go:
```diff
@@ -20,8 +19,8 @@ const (
 	ErrQueuedTxExist = errors.New("transaction already exist in queue")
-	//ErrQueuedTxIDNotFound - error transaction hash not found
-	ErrQueuedTxIDNotFound = errors.New("transaction hash not found")
+	//ErrstringNotFound - error transaction hash not found
```
Find&Replace did not work well here 😅

adriacidre replied:
OMG! 😅

adambabik commented on geth/transactions/utils.go:
```diff
@@ -23,23 +21,24 @@ const (
-		os.MkdirAll(keystoreDir, os.ModePerm) // nolint: errcheck, gas
-	}
+// HaltOnPanic recovers from panic, logs issue, sends upward notification, and exits
+func HaltOnPanic() {
```
Do we need to export it? I would love to get rid of this function in the future so let's avoid it's used anywhere else.

adriacidre replied:
Nope, it's only used inside transactions package, I've also made private `MessageIDFromContext` and `Fatalf` as they were only used on transactions package too.

adambabik commented on t/e2e/jail/jail_rpc_test.go:
```diff
@@ -291,7 +290,7 @@ func (s *JailRPCTestSuite) TestJailVMPersistence() {
 			//var txHash common.Hash
 			txID := event["id"].(string)
-			txHash, e := s.Backend.CompleteTransaction(common.QueuedTxID(txID), TestConfig.Account1.Password)
+			txHash, e := s.Backend.CompleteTransaction(string(txID), TestConfig.Account1.Password)
```
`txID` is already a string.

adambabik commented on t/e2e/jail/jail_rpc_test.go:
```diff
@@ -135,7 +134,7 @@ func (s *JailRPCTestSuite) TestContractDeployment() {
 			txID := event["id"].(string)
 			var txErr error
-			txHash, txErr = s.Backend.CompleteTransaction(common.QueuedTxID(txID), TestConfig.Account1.Password)
+			txHash, txErr = s.Backend.CompleteTransaction(string(txID), TestConfig.Account1.Password)
```
`txID` is already a string.

divan commented on geth/api/backend.go:
```diff
@@ -193,11 +192,11 @@ func (b *StatusBackend) CallRPC(inputJSON string) string {
 		ctx = context.Background()
 	}
-	tx := common.CreateTransaction(ctx, args)
+	tx := transactions.CreateTransaction(ctx, args)
```
Let's double check exported methods and variables naming after moving from `common`.
For example, `transactions.CreateTransaction()` should look much clearer as `transactions.Create()`.

adriacidre replied:
@divan simplified exported methods I've moved from `common` package. 

There is also some margin for improvement on pre-existing transactions methods/vars naming, but imho, may be better to refactor them on a different iteration.

mandrigin replied:
Shouldn't it be `New` then?
`errors.New("blah")`, `transactions.New("blah")` looks more consistent with stdlib.

adriacidre replied:
Yep, but since we moved `transactions/queue` package inside `transactions` this package is managing different "objects":

`transactions.New` results on a new `TxQueue`
while:
`transactions.NewManager` on a new `Manager`

adriacidre replied:
hmmm... as far as i can see `transactions.New` is not being used outside of transactions package. I'll change that function name to something like `newQueue` and rename `transactions.NewManager` to `transactions.New` :)

divan reviewed (changes requested):
Fantastic job. I've been dreaming about ditching `common` package for a long time!

So it worth to review the naming after moving things from `common` to `transactions` – there is a lot of room to make it much clearer, like `transactions.DiscardTransactionResult()` -> `transactions.DiscardResult()`.

adriacidre commented:
@divan, @adambabik, @mandrigin all review changes implemented!

adambabik commented on geth/transactions/txqueue_manager.go:
```diff
@@ -293,13 +292,13 @@ func (m *Manager) DiscardTransaction(id common.QueuedTxID) error {
 // DiscardTransactions discards given multiple transactions from transaction queue
-func (m *Manager) DiscardTransactions(ids []common.QueuedTxID) map[common.QueuedTxID]common.RawDiscardTransactionResult {
-	results := make(map[common.QueuedTxID]common.RawDiscardTransactionResult)
+func (m *Manager) DiscardTransactions(ids []string) map[string]RawDiscardResult {
```
We can get rid of this method. There is no `CompleteTransactions`, only `CompleteTransaction`.

adriacidre replied:
That's been used on the [backend](https://github.com/status-im/status-go/blob/0e05743109560ce60b8b95ccad80669a84b55be3/geth/api/backend.go#L259)

adambabik commented on geth/transactions/types.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+// RawDiscardResult is list of results from CompleteTransactions() (used internally)
+type RawDiscardResult struct {
```
Is this type needed for anything?

adriacidre replied:
Used as return type on the [backend](https://github.com/status-im/status-go/blob/0e05743109560ce60b8b95ccad80669a84b55be3/geth/api/backend.go#L259) and [api](https://github.com/status-im/status-go/blob/0e05743109560ce60b8b95ccad80669a84b55be3/geth/api/api.go#L169)

adambabik replied:
Yeah, but it looks like we can replace it with `error` and it will work the same, right?

adriacidre replied:
yep, you're right :)

adambabik reviewed (approved):
LGTM Great stuff as always 👏

divan commented on geth/api/backend.go:
```diff
@@ -49,7 +48,7 @@ func NewStatusBackend() *StatusBackend {
 	nodeManager := node.NewManager()
 	accountManager := account.NewManager(nodeManager)
-	txQueueManager := transactions.NewManager(nodeManager)
+	txQueueManager := transactions.New(nodeManager)
```
`transactions.New()` actually suggests that it's a new transaction is being created. I would keep `NewManager()` for this specific case.

divan reviewed (approved):
+421 - 542 🥇 Great job!

divan commented on geth/transactions/txqueue_manager.go:
```diff
@@ -48,11 +47,11 @@ type Manager struct {
 }
 
-// NewManager returns a new Manager.
+// New returns a new Manager.
```
forgot comment :)

divan commented on geth/transactions/txqueue_manager.go:
```diff
@@ -48,11 +47,11 @@ type Manager struct {
 }
 
-// NewManager returns a new Manager.
+// New returns a new Manager.
```
forgot comment :)

adriacidre replied:
ups

adriacidre commented:
all done! i'll merge once CI is green, cheers for your comments :)
